File: Algorithms/Data Structure dependent/Arrays and String/regex/validNumber.py
class Solution:
    # (\+|-) - handles optional sign
    # ((\d+(\.)?\d*) - handles both integer an decimals with dot at end or 
    # decimal with dot on both sides
    # (\.\d+) - handles final decimal cases
    # ((e|E)(\+|-)?\d+)? - handles optional exponent 
    # NOTE: we could just use a single "e" and ignore case here as well 
    
    reg = re.compile(r"^((\+|-)?((\d+(\.)?\d*)|(\.\d+))((e|E)(\+|-)?\d+)?)$")
    def isNumber(self, s: str) -> bool:
        return self.reg.fullmatch(s)
        

File: Algorithms/Data Structure dependent/Arrays and String/regex/validNumber.cpp
class Solution {
public:
    bool isNumber(string s) {
        /* (\+|-)? - Start with optional sign
         * Now we have 2 possible cases
         * 1. (\.\d+)  - Decimal with dot at front
         * 2. ((\d+)(\.)?\d*)  - Any other decimal or integer
         * End with
         * ((E|e)(+|-)?(\d)+)? - Allow Exponent
         */
        const static std::regex re(R"((\+|-)?((\.\d+)|((\d+)(\.)?\d*))((E|e)(\+|-)?(\d)+)?)");
        return std::regex_match(s, re);
    }
};